lambdaclass/champions_of_mirra
# [GH-1617] connectivity issues feedback

> [!WARNING]  
> This has to be merged with https://github.com/lambdaclass/mirra_backend/pull/619

Works towards solving #1617 

## Motivation

When a user has connectivity issues, feedback communicating this should be returned to the user.

## Summary of changes

- An algorithm measuring jitter (high variations of ping between messages) and the time elapsed since the last update were implemented.
- When any of these two cases are met, a bad connection icon will be display on the center of the screen.
- For the icon to disappear the connection must be stabilized.
- If the client receives no messages for 10 seconds, then it will return to the main screen with a disconnection message popup.

All the parameters used in the algorithms to detect the stability can be set in the backend file `client_config.json`.

## How has this been tested?

### To test the complete lack of messages

This won't work if you are testing on the Unity editor since to disconnect the computer to the internet or kill the backend running locally you have to loose focus of the application, and by the way Unity handles this the icon will appear, even if no connection issues arise, tainting the results of the test.

- Enter a match and in the middle of it disconnect your device from the backend. You should see the icon popup after 3 seconds without internet, then after 10 seconds the game should take you to the main screen with a popup explaining what happened.

### To test jittering

- You'll have to try this al least once with a good connection to see that the icon doesn't appear.
- After that try to play a match with a bad connection, see that when the match itself feels laggy the icon is displayed, and when the connection is stable again, and you can play correctly, the icon disappears. It's best to pay attention to the bots in the game to see if your connection is good, since your own character has client prediction so it might move correctly even with a bad connection.

## Checklist
- [x] I have tested the whole game after applying the changes, not only the affected areas.
